[Docs] refactor get_started.md (#457)
open-mmlab/mmselfsup: docs/en/get_started.md

- [Get Started](#get-started)
  - [Prerequisites](#prerequisites)
  - [Installation](#installation)
    - [Best practices](#best-practices)
    - [Verify the installation](#verify-the-installation)
    - [Customize installation](#customize-installation)
      - [Benchmark](#benchmark)
      - [CUDA versions](#cuda-versions)
      - [Install MMEngine without MIM](#install-mmengine-without-mim)
      - [Install MMCV without MIM](#install-mmcv-without-mim)
      - [Install on CPU-only platforms](#install-on-cpu-only-platforms)
      - [Install on Google Colab](#install-on-google-colab)
      - [Using MMSelfSup with Docker](#using-mmselfsup-with-docker)
    - [Trouble shooting](#trouble-shooting)
  - [Using Multiple MMSelfSup Versions](#using-multiple-mmselfsup-versions)

## Prerequisites

In this section, we demonstrate how to prepare an environment with PyTorch.

> MMSelfSup works on Linux (Windows and macOS are not officially supported). It requires Python 3.6+, CUDA 9.2+ and PyTorch 1.5+.

Reviewer: ```suggestion
MMSelfSup works on Linux (Windows and macOS are not officially supported). It requires Python 3.6+, PyTorch 1.6+.
```

```{note}

> If you are experienced with PyTorch and have already installed it, just skip this part and jump to the [next section](#installation). Otherwise, you can follow these steps for the preparation.

Reviewer 2: It's not a link.
```{note}
[next section](#installation)
```

## Installation

We recommend users to follow our best practices to install MMSelfSup. However, the whole process is highly customizable. See [Customize Installation](#customize-installation) section for more information.

### Best practices

**Step 0.** Install [MMEngine](https://github.com/open-mmlab/mmengine) and [MMCV](https://github.com/open-mmlab/mmcv) using [MIM](https://github.com/open-mmlab/mim).

```shell
pip install -U openmim

> mim install mmengine "mmcv-full>=2.0rc0"

Reviewer: ```suggestion
mim install mmengine "mmcv>=2.0.0rc1"
```

```

**Step 1.** Install MMSelfSup.

According to your needs, we support two installation modes:

- [Install from source (Recommended)](#install-from-source): You want to develop your own self-supervised task or new features based on MMSelfSup framework, e.g., adding new datasets or models. And you can use all tools we provided.
- [Install as a Python package](#install-as-a-python-package): You just want to call MMSelfSup's APIs or import MMSelfSup's modules in your project.

#### Install from source

In this case, install mmselfsup from source:

```shell
git clone https://github.com/open-mmlab/mmselfsup.git
cd mmselfsup
git checkout 1.x
pip install -v -e .
# "-v" means verbose, or more output
# "-e" means installing a project in editable mode,
# thus any local modifications made to the code will take effect without reinstallation.
```

Optionally, if you want to contribute to MMSelfSup or experience experimental functions, please checkout to the `dev-1.x` branch:

```shell
git checkout dev-1.x
```

#### Install as a Python package

Just install with pip.

```shell

> pip install "mmselfsup>=1.0rc0"

Reviewer: ```suggestion
pip install "mmselfsup>=1.0.0rc0"
```

```

### Verify the installation

To verify whether MMSelfSup is installed correctly, you can run the following sample code to initialize a model and inference a demo image.

```python

> import torch

Reviewer: i think we could use version check instead of it
E.g.
```python
import mmselfsup
print(mmselfsup.__version__)
```

The above code is supposed to run successfully upon you finish the installation.

### Customize installation

#### Benchmark

The [Best practices](#best-practices) is for basic usage. If you need to evaluate your pre-trained model with some downstream tasks such as detection or segmentation, please also install [MMDetection](https://github.com/open-mmlab/mmdetection) and [MMSegmentation](https://github.com/open-mmlab/mmsegmentation).

If you don't run MMDetection and MMSegmentation benchmarks, it is unnecessary to install them.

You can simply install MMDetection and MMSegmentation with the following command:

```shell

> pip install mmdet mmsegmentation

Reviewer: ```suggestion
pip install 'mmdet>=3.0.0rc0' 'mmsegmentation>=1.0.0rc0'
```

```

#### Install MMEngine without MIM

To install MMEngine with pip instead of MIM, please follow [MMEngine installation guides](https://mmcv.readthedocs.io/en/latest/get_started/installation.html).

For example, you can install MMEngine by the following command.

```shell
pip install mmengine
```

#### Install MMCV without MIM

MMCV contains C++ and CUDA extensions, thus depending on PyTorch in a complex way. MIM solves such dependencies automatically and makes the installation easier. However, it is not a must.

To install MMCV with pip instead of MIM, please follow [MMCV installation guides](https://mmcv.readthedocs.io/en/latest/get_started/installation.html). This requires manually specifying a find-url based on PyTorch version and its CUDA version.

For example, the following command install mmcv-full built for PyTorch 1.10.x and CUDA 11.3.

```shell

> pip install "mmcv-full>=2.0rc0" -f https://download.openmmlab.com/mmcv/dist/cu113/torch1.10/index.html

Reviewer: ```suggestion
pip install "mmcv-full>=2.0.0rc1" -f https://download.openmmlab.com/mmcv/dist/cu113/torch1.10/index.html
```

```

### Install on CPU-only platforms

MMSelfSup can be built for CPU only environment. In CPU mode, you can train, test or inference a model.

Some functionalities are gone in this mode, usually GPU-compiled ops. But don't
worry, almost all models in MMClassification don't depends on these ops.

#### Install on Google Colab

[Google Colab](https://research.google.com/) usually has PyTorch installed,
thus we only need to install MMCV and MMSeflSup with the following commands.

**Step 0.** Install [MMEngine](https://github.com/open-mmlab/mmengine) and [MMCV](https://github.com/open-mmlab/mmcv) using [MIM](https://github.com/open-mmlab/mim).

```shell
!pip3 install openmim

> !mim install mmengine "mmcv-full>=2.0rc0"

Reviewer: ```suggestion
!mim install mmengine "mmcv>=2.0.0rc0"
```

Installing CUDA runtime libraries is enough if you follow our best practices, because no CUDA code will be compiled locally. However if you hope to compile MMCV from source or develop other CUDA operators, you need to install the complete CUDA toolkit from NVIDIA's [website](https://developer.nvidia.com/cuda-downloads), and its version should match the CUDA version of PyTorch. i.e., the specified version of cudatoolkit in `conda install` command.
```

#### Install MMEngine without MIM

> To install MMEngine with pip instead of MIM, please follow [MMEngine installation guides](https://mmcv.readthedocs.io/en/latest/get_started/installation.html).

Reviewer: ```suggestion
To install MMEngine with pip instead of MIM, please follow [MMEngine installation guides](https://github.com/open-mmlab/mmengine/blob/main/docs/zh_cn/get_started/installation.md).
```

For example, you can install MMEngine by the following command.

```shell
pip install mmengine
```

#### Install MMCV without MIM

MMCV contains C++ and CUDA extensions, thus depending on PyTorch in a complex way. MIM solves such dependencies automatically and makes the installation easier. However, it is not a must.

> To install MMCV with pip instead of MIM, please follow [MMCV installation guides](https://mmcv.readthedocs.io/en/latest/get_started/installation.html). This requires manually specifying a find-url based on PyTorch version and its CUDA version.

Reviewer: ```suggestion
To install MMCV with pip instead of MIM, please follow [MMCV installation guides](https://mmcv.readthedocs.io/en/2.x/get_started/installation.html). This requires manually specifying a find-url based on PyTorch version and its CUDA version.
```

> For example, the following command install mmcv-full built for PyTorch 1.10.x and CUDA 11.3.

Reviewer: update version as mentioned in feishu group

```shell

> pip install mmcv-full>=2.0.0rc1 -f https://download.openmmlab.com/mmcv/dist/cu113/torch1.10/index.html

Reviewer: ```suggestion
pip install mmcv-full>=2.0.0rc1 -f https://download.openmmlab.com/mmcv/dist/cu116/torch1.12.0/index.html
```

```

### Install on CPU-only platforms

MMSelfSup can be built for CPU only environment. In CPU mode, you can train, test or inference a model.

Some functionalities are gone in this mode, usually GPU-compiled ops. But don't

> worry, almost all models in MMClassification don't depends on these ops.

Reviewer 2: ```suggestion
worry, almost all models in MMSelfsup don't depends on these ops.
```

```python
import mmselfsup
print(mmselfsup.__version__)
# Example output: 1.0.0rc0 or newer
```

```{note}
Within Jupyter, the exclamation mark `!` is used to call external executables and `%cd` is a [magic command](https://ipython.readthedocs.io/en/stable/interactive/magics.html#magic-cd) to change the current working directory of Python.
```

#### Using MMSelfSup with Docker

We provide a [Dockerfile](https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/docker/Dockerfile) to build an image. Ensure that your [docker version](https://docs.docker.com/engine/install/) >=19.03.

```shell
# build an image with PyTorch 1.10.0, CUDA 11.3, CUDNN 8.
docker build -f ./docker/Dockerfile --rm -t mmselfsup:torch1.10.0-cuda11.3-cudnn8 .
```

**Important:** Make sure you've installed the [nvidia-container-toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#docker).

Run the following command:

```shell
docker run --gpus all --shm-size=8g -it -v {DATA_DIR}:/workspace/mmselfsup/data mmselfsup:torch1.10.0-cuda11.3-cudnn8 /bin/bash
```

`{DATA_DIR}` is your local folder containing all these datasets.

### Trouble shooting

> If you have some issues during the installation, please first view the [FAQ](https://github.com/open-mmlab/mmselfsup/blob/dev-1.x/docs/en/notes/faq.md) page.

Reviewer: ```suggestion
If you have some issues during the installation, please first view the [FAQ](notes/faq.md) page.
```
